token/program-2022/src/instruction.rs

```diff
@@ -409,6 +427,126 @@ pub enum TokenInstruction {
         /// The freeze authority/multisignature of the mint.
         freeze_authority: COption<Pubkey>,
     },
+    /// Initialize a new mint with all provided mixins.
+    ///
+    /// The mint must have exactly enough space allocated for the base mint (82
+    /// bytes). If any mixins are provided, there must be 83 bytes of padding,
+    /// 1 byte reserved for the account type, then space required by each mixin.
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   0. `[writable]` The mint to initialize.
+    ///   1. ..1+N `[]` (Optional) Accounts required to initialize the mixins
+    InitializeMintWithMixins {
```

Comment:
What if we only had `InitializeAccount`/`InitializeMint` still.  When the user wants to adds mixins then they just over-allocate additional account data

Each mixin then has its own initializer, like `InitializeTransferFee`, that follows `InitializeMint`.

Interestingly it seems that an Account doesn't even need an explicit `InitializeTransferFee`.  The Mint is passed into `InitializeAccount`, so if the Mint has a transfer fee initialized then the token program ought to initialize it implicitly for the Account (or fail if the provided account doesn't have enough data storage for the transfer fee mixin)

Reply:
> What if we only had InitializeAccount/InitializeMint still. When the user wants to adds mixins then they just over-allocate additional account data

That was my initial idea, but then I was worried about ensuring that everything is done in one transaction.  If an account requires `InitializeTransferFee`, how are we sure that it's done in the same transaction?  That requires `InitializeMint` to take in the instructions sysvar and check that the correct accounts are passed in, which adds more compute unit usage and complexity.  I worry about undefined behavior / attack opportunities otherwise.  The alternative is to have `InitializeMint` done last, and mixin initializers (ie. `InitializeTransferFee`) fail if any data has been written in the `Mint` part.  But then I worry that it might not be usable if any data from the `Mint` is required when initializing a mixin.

> Interestingly it seems that an Account doesn't even need an explicit InitializeTransferFee. The Mint is passed into InitializeAccount, so if the Mint has a transfer fee initialized then the token program ought to initialize it implicitly for the Account (or fail if the provided account doesn't have enough data storage for the transfer fee mixin)

Absolutely! This was also my initial idea, but you mentioned earlier that you wanted to be explicit about everything, so I put in the explicit instruction.  I'd prefer to just use `InitializeAccount` and a helper function on the mint to give the size of an Account.  That way, things like ATA need only tiny changes.

Reply:
> but then I was worried about ensuring that everything is done in one transaction 
> ... 
> any data from the Mint is required when initializing a mixin.

Ok I'm convinced to keep it as for now and we can see how the fully implementation feels

>  I'd prefer to just use InitializeAccount and a helper function on the mint to give the size of an Account. That way, things like ATA need only tiny changes.

Heh, ok deal.  I do think there could be future mixins that do still require Account-level initialization of course